### PR TITLE
test: add inline event assertions for resource template events

### PR DESCRIPTION
**Pull request by nXtCyberNet**
**What type of PR is this?**

/kind feature
/kind test

---

**What this PR does / why we need it**:

This PR adds inline event assertions to existing E2E test cases to improve event coverage for resource template events, as discussed in the related issue.

Following the community decision to use **Approach 2 (Inline Event Assertions)**, event checks are added directly inside existing test cases at the point where the relevant operation completes — no new test cases or files are introduced.

### `test/e2e/framework/events.go`

* Added `WaitEventV1FitWith` helper to support `events.k8s.io/v1` Event API alongside the existing `core/v1` helper (foundation for future migration)

### `test/e2e/suites/base/propagationpolicy_test.go`

Added inline event assertions to existing Deployment, Service, and Job propagation test cases:

* `ApplyPolicySucceed` — verified on Deployment, Service, Job
* `GetReplicasSucceed` — verified on Deployment, Job
* `ScheduleBindingSucceed` — verified on Deployment, Service, Job
* `SyncWorkSucceed` — verified on Deployment, Service, Job
* `AggregateStatusSucceed` — verified on Deployment, Job
* `ReflectStatusSucceed` — verified on Work object during Deployment propagation
* `InterpretHealthSucceed` — verified on Work object during Deployment propagation
* `SyncWorkloadSucceed` — verified on Work object during Deployment propagation

### `test/e2e/suites/base/overridepolicy_test.go`

Added inline event assertions to existing override policy test cases:

* `ApplyOverridePolicySucceed` — verified across OverridePolicy and OverrideRules test cases with Deployment

### `test/e2e/suites/base/preemption_test.go`

Added inline event assertions to existing preemption test cases:

* `PreemptPolicySucceed` — verified when high-priority PropagationPolicy preempts low-priority PropagationPolicy, when PropagationPolicy preempts ClusterPropagationPolicy, and when high-priority ClusterPropagationPolicy preempts low-priority ClusterPropagationPolicy

### `test/e2e/suites/base/dependenciesdistributor_test.go`

Added inline event assertions to existing dependency propagation test cases:

* `GetDependenciesSucceed` — verified in configmap automatic propagation tests with both PropagationPolicy and ClusterPropagationPolicy

---

**Which issue(s) this PR fixes**:

Fixes part of  #7252

---

**Special notes for your reviewer**:

## How was this tested

All assertions were verified by running the E2E suite locally against a kind-based Karmada setup and confirming the expected events fire during the propagation flow.

```bash
export KUBECONFIG=$HOME/.kube/karmada.config
go test ./test/e2e/suites/base/... \
  -v \
  --ginkgo.focus="BasicCase|OverridePolicy|OverrideRules|Preemption|DependenciesDistributor" \
  --ginkgo.timeout=60m \
  -- \
  --host-context=karmada-host \
  --karmada-context=karmada-apiserver \
  --poll-timeout=300s
```

## Notes

* `GetDependenciesFailed`, `GetReplicasFailed`, and `ApplyPolicyFailed` are not covered in this PR as they require specific failure conditions that are difficult to construct reliably in E2E tests
* `events.k8s.io/v1` support added in `framework/events.go` is not yet used in assertions — it is foundation work for the planned migration from `GetEventRecorderFor` to `GetEventRecorder`

---

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/framework/events.go [hunk 1]
```diff
@@ -37,3 +38,18 @@ func WaitEventFitWith(kubeClient kubernetes.Interface, namespace string, involve
...
+		eventList, err := kubeClient.EventsV1().Events(namespace).List(context.TODO(), metav1.ListOptions{
+			FieldSelector: fields.OneTermEqualSelector("regarding.name", involvedObj).String(),
+		})
+		gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+		for _, e := range eventList.Items {
+			if fit(e) {
+				return true
+			}
+		}
+		return false
```
[the remark this answers is not in the document]
Comment (nXtCyberNet) [reply test/e2e/framework/events.go]: its  because the k8s/event/v1 will use regarding.name instead of involvedObject.name

[the remark this answers is not in the document]
Comment (nXtCyberNet) [reply test/e2e/framework/events.go]: events.k8s.io/v1 events are best-effort and field selector behavior can be inconsistent, so explicit iteration is preferred for better debugging and reliability in E2E tests

**Comment by XiShanYongYe-Chang** [comment]
> Fixes https://github.com/karmada-io/karmada/issues/7252

This will close the target issue. Please adjust the wording: Fixes part of xx

File: test/e2e/framework/events.go [hunk 2]
```diff
@@ -37,3 +38,20 @@ func WaitEventFitWith(kubeClient kubernetes.Interface, namespace string, involve
 		return slices.ContainsFunc(eventList.Items, fit)
 	}, PollTimeout, PollInterval).Should(gomega.Equal(true))
 }
+
+// WaitEventV1FitWith uses events.k8s.io/v1 to do the same thing as WaitEventFitWith
+func WaitEventV1FitWith(kubeClient kubernetes.Interface, namespace string, involvedObj string, fit func(eventsv1.Event) bool) {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/framework/events.go]: Support for eventsv1 can be added later.

**Comment by XiShanYongYe-Chang** [comment]
Once this PR is ready, you can squash all the commits into one and then ask me to review it.

File: test/e2e/framework/events.go (hunk 2 above)
Comment (nXtCyberNet) [reply test/e2e/framework/events.go]: Okayy , I will remove it

Comment (nXtCyberNet) [reply test/e2e/framework/events.go]: @XiShanYongYe-Chang  btw I didn’t use slices.ContainsFunc here because of the somewhat unpredictable nature of the events/v1 API (ordering, transient updates, etc.). does this issue still exist in newer versions, or is it safe to use ContainsFunc now? just asking for my knowledge, thanks.

Comment (XiShanYongYe-Chang) [reply test/e2e/framework/events.go]: I don't quite understand why we can't use ContainsFunc.

**Comment by nXtCyberNet** [comment]
> Once this PR is ready, you can squash all the commits into one and then ask me to review it.

Hi, I’ve squashed all the commits into one. The PR is ready for review, please take a look

**Comment by XiShanYongYe-Chang** [comment]
Thanks
/assign

File: test/e2e/suites/base/overridepolicy_test.go [hunk 3]
```diff
@@ -121,6 +122,12 @@ var _ = ginkgo.Describe("[OverridePolicy] apply overriders testing", func() {
 						deployment.GetLabels()["foo"] == "exist" &&
 						deployment.GetLabels()["app"] == "nginx"
 				})
+			ginkgo.By("checking ApplyOverridePolicySucceed event on deployment", func() {
+				framework.WaitEventFitWith(kubeClient, deploymentNamespace, deploymentName,
+					func(event corev1.Event) bool {
+						return event.Reason == events.EventReasonApplyOverridePolicySucceed
+					})
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/overridepolicy_test.go]: e2e runs in parallel, and events exist for a period of time after being created. If we only check the reason without checking the corresponding resource name, can we effectively make a judgment?

Comment (nXtCyberNet) [reply test/e2e/suites/base/overridepolicy_test.go]: Hi, I believe this should not be an issue since all resources are created in their respective namespaces. Each namespace, along with its deployments, follows a consistent naming pattern, ensuring proper isolation.
Additionally, we are matching against InvolvedObject.Name, which is sufficiently unique, reducing the likelihood of false positive signals from unrelated resources.
That said, please let me know if I might be missing any edge cases or scenarios that should be considered.

Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/overridepolicy_test.go]: Yes, you are right

File: test/e2e/framework/events.go (hunk 2 above)
Comment (XiShanYongYe-Chang) [inline test/e2e/framework/events.go]: Delete this function first?

Comment (nXtCyberNet) [reply test/e2e/framework/events.go]: Apologies, I overlooked that earlier. I’ve included it now

File: test/e2e/suites/base/dependenciesdistributor_test.go [hunk 4]
```diff
@@ -113,6 +113,12 @@ var _ = ginkgo.Describe("[DependenciesDistributor] automatically propagate relev
 							return true
 						})
 				})
+				ginkgo.By("checking GetDependenciesSucceed event on deployment", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/dependenciesdistributor_test.go:116]: There are also dependency distribution scenarios for other resources in this test file. Why were these not tested together?

File: test/e2e/suites/base/overridepolicy_test.go [hunk 5]
```diff
@@ -121,6 +122,12 @@ var _ = ginkgo.Describe("[OverridePolicy] apply overriders testing", func() {
 						deployment.GetLabels()["foo"] == "exist" &&
 						deployment.GetLabels()["app"] == "nginx"
 				})
+			ginkgo.By("checking ApplyOverridePolicySucceed event on deployment", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/overridepolicy_test.go]: The check on an event can be moved before the check on resource distribution, as shown in line 116.

Can the same test be performed on ClusterOverridePolicy?

File: test/e2e/suites/base/preemption_test.go [hunk 6]
```diff
@@ -87,6 +89,12 @@ var _ = ginkgo.Describe("[Preemption] propagation policy preemption testing", fu
 					framework.CreatePropagationPolicy(karmadaClient, highPriorityPolicy)
 					framework.WaitDeploymentPresentOnClusterFitWith(preemptingClusterName, deployment.Namespace, deployment.Name, func(*appsv1.Deployment) bool { return true })
 				})
+				ginkgo.By("checking PreemptPolicySucceed event on deployment", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/preemption_test.go:92]: Similarly, I think event checks can be moved before resource checks.

File: test/e2e/suites/base/propagationpolicy_test.go [hunk 7]
```diff
@@ -157,59 +226,50 @@ var _ = ginkgo.Describe("[BasicCase] PropagationPolicy testing", func() {
 					return true
 				})
 
-			patch := []map[string]any{{"op": "replace", "path": "/spec/ports/0/port", "value": updateServicePort}}
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/propagationpolicy_test.go:160]: Why do we need to remove these codes? Same as below.

File: test/e2e/suites/base/propagationpolicy_test.go [hunk 8]
```diff
@@ -97,16 +97,85 @@ var _ = ginkgo.Describe("[BasicCase] PropagationPolicy testing", func() {
...
+					})
+			})
 
-			framework.UpdateDeploymentReplicas(kubeClient, deployment, updateDeploymentReplicas)
-			framework.WaitDeploymentPresentOnClustersFitWith(framework.ClusterNames(), deployment.Namespace, deployment.Name,
-				func(deployment *appsv1.Deployment) bool {
-					return *deployment.Spec.Replicas == updateDeploymentReplicas
-				})
+			// resource template events — fired on the Deployment object itself
+			ginkgo.By("checking ApplyPolicySucceed event", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/propagationpolicy_test.go:100]: I think the event check here can be divided into two parts. The first part is the event for resource distribution, which can be checked before WaitDeploymentPresentOnClustersFitWith. The second part is the event for status collection, which can be checked after WaitDeploymentPresentOnClustersFitWith. WDYT

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks a lot

**Comment by XiShanYongYe-Chang** [comment]
Hi @nXtCyberNet The first PR may have a lot of reviews, but subsequent PRs should be much better if they are based on this template.

File: test/e2e/suites/base/overridepolicy_test.go (hunk 5 above)
Comment (nXtCyberNet) [reply test/e2e/suites/base/overridepolicy_test.go]: Hi @XiShanYongYe-Chang, thank you for the feedback. I’ll move the event assertion before the resource distribution check in all affected It blocks. Currently, this PR is focused on the resource template subtask; I’ll add ApplyOverridePolicySucceed assertions to the ClusterOverridePolicy tests in another PR

File: test/e2e/suites/base/dependenciesdistributor_test.go (hunk 4 above)
Comment (nXtCyberNet) [reply test/e2e/suites/base/dependenciesdistributor_test.go:116]: You're right — those scenarios were missed earlier as each When block creates isolated deployments; I'll  added the dependency propagation assertions for all remaining resources.

File: test/e2e/suites/base/propagationpolicy_test.go (hunk 8 above)
Comment (nXtCyberNet) [reply test/e2e/suites/base/propagationpolicy_test.go:100]: One concern I want to raise before finalizing the grouping approach:

Since Kubernetes events are best-effort and not guaranteed to be delivered (they can be dropped under memory pressure or aggregated), is there a risk that WaitEventFitWith could cause a flaky test failure simply because an event was dropped rather than because the feature actually broke?

I noticed the existing event assertions in the codebase (e.g. WorkDispatching, DependencyPolicyConflict) make the same assumption. Just wanted to confirm this tradeoff is acceptable for this PR as well before proceeding — I can adjust the approach if needed.

Comment (nXtCyberNet) [reply test/e2e/suites/base/propagationpolicy_test.go:100]: One follow-up thought — if events are dropped, wouldn’t the existing By block checks also fail for the same reason? In that case, the flakiness risk isn’t specific to WaitEventFitWith but applies to the current approach as well.

Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/propagationpolicy_test.go:100]: Very good question.

>  I can adjust the approach if needed.

Do you have any good ideas?

Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/propagationpolicy_test.go:100]: > One follow-up thought — if events are dropped, wouldn’t the existing By block checks also fail for the same reason? In that case, the flakiness risk isn’t specific to WaitEventFitWith but applies to the current approach as well.

@gemini-code-assist
Do you have any good suggestions?

Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/propagationpolicy_test.go:100]: Based on the above analysis, I think that our current configuration with a longer timeout should be acceptable. If we find out later that events are being dropped, then we can consider the `Non-blocking checks:` again. WDYT @nXtCyberNet

Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/propagationpolicy_test.go:100]: Kindly ping @nXtCyberNet 
Can we adjust the order of the inspection based on the first comments?

File: test/e2e/suites/base/overridepolicy_test.go [hunk 9]
```diff
@@ -113,6 +114,12 @@ var _ = ginkgo.Describe("[OverridePolicy] apply overriders testing", func() {
 
 		ginkgo.It("deployment labelsOverride testing", func() {
 			klog.Infof("check if deployment present on member clusters have correct labels value")
+			ginkgo.By("checking ApplyOverridePolicySucceed event on deployment", func() {
+				framework.WaitEventFitWith(kubeClient, deploymentNamespace, deploymentName,
+					func(event corev1.Event) bool {
+						return event.Reason == events.EventReasonApplyOverridePolicySucceed
+					})
+			})
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/overridepolicy_test.go:121]: If this logic is moved to the previous line of log, it would feel better.

File: pkg/events/events.go
Comment (XiShanYongYe-Chang) [inline pkg/events/events.go:134]: Why do we need to add some new events for the current PR?

Comment (nXtCyberNet) [reply pkg/events/events.go]: You're right, GetComponentsSucceed and GetComponentsFailed are out of scope for this PR. I noticed those event emissions were missing while working on the detector, but they aren't needed for the current assertions. I'll remove them and can submit them separately if needed.

File: test/e2e/suites/base/propagationpolicy_test.go [hunk 10]
```diff
@@ -167,53 +274,6 @@ var _ = ginkgo.Describe("[BasicCase] PropagationPolicy testing", func() {
 		})
 	})
 
-	ginkgo.Context("Pod propagation testing", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/suites/base/propagationpolicy_test.go:170]: Was this test mistakenly deleted?

other parts lgtm

Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/propagationpolicy_test.go:170]: Could you please help remind me of any changes?

**Comment by mohamedawnallah** [comment]
Thanks @nXtCyberNet for the PR!

I ran mutation testing on the line of code this PR exercises: https://github.com/karmada-io/karmada/blob/b15f5bd70ed3cf6e62bbea9126e1e1053d979030/pkg/detector/detector.go#L540

Happy path: Ran tests against the existing code — `ApplyPolicySucceed` passed in ~55 seconds.

Mutation test: Removed the target line from the source code so the test should fail. The mutation run failed at exactly 420.002 seconds (420s hardcoded `framework.PollTimeout`) on the `ApplyPolicySucceed` event check step.

Nice work. LGTM pending XiShanYongYe-Chang's comment: https://github.com/karmada-io/karmada/pull/7300#discussion_r3055197680

**Comment by nXtCyberNet** [comment]
Hi @mohamedawnallah , thank you so much for the kind words and for taking the time to run mutation testing on the PR — that level of thoroughness is really appreciated and something I'll keep in mind for future contributions. Looking forward to making more meaningful PRs to the repo! 🙌

File: test/e2e/suites/base/propagationpolicy_test.go (hunk 10 above)
Comment (XiShanYongYe-Chang) [reply test/e2e/suites/base/propagationpolicy_test.go:170]: Hi @nXtCyberNet This use case was still wrongly deleted; we need keep it.

**Comment by mohamedawnallah** [comment]
> Hi @mohamedawnallah , thank you so much for the kind words and for taking the time to run mutation testing on the PR — that level of thoroughness is really appreciated and something I'll keep in mind for future contributions. Looking forward to making more meaningful PRs to the repo! 🙌

Thanks @nXtCyberNet, PTAL at the comment again https://github.com/karmada-io/karmada/pull/7300#discussion_r3071818720. The PR merging is pending addressing that comment. Thanks

**Comment by nXtCyberNet** [comment]
Hi, I believe the CI failures on Kubernetes v1.35.0 are due to environment instability rather than a code regression — I've opened issue #7388 to track this. If both of you agree, I'd love to investigate the flakiness further and work on a fix. Thanks!

File: test/e2e/suites/base/propagationpolicy_test.go (hunk 10 above)
Comment (nXtCyberNet) [reply test/e2e/suites/base/propagationpolicy_test.go]: sorry for the misunderstanding i have added that test  again!

**Comment by mohamedawnallah** [comment]
Thanks @nXtCyberNet for addressing comment and having tracking issue documenting the observed flaky issues. It LGTM, just leaving space for @XiShanYongYe-Chang for any comments may have

**Comment by XiShanYongYe-Chang** [comment]
Hi @nXtCyberNet can you help squash the commits into one?

**Comment by mohamedawnallah** [comment]
@XiShanYongYe-Chang, I see @nXtCyberNet squashed the commits. Are we ready to merge?

**Comment by XiShanYongYe-Chang** [comment]
Thanks for reminding @mohamedawnallah 
/approve

**Comment by XiShanYongYe-Chang** [comment]
/retest
Leave lgtm to @mohamedawnallah for more check.
